builtin/logical/database/rotation.go

```diff
@@ -330,61 +351,124 @@ func (b *databaseBackend) setStaticAccount(ctx context.Context, s logical.Storag
 	dbi.RLock()
 	defer dbi.RUnlock()
 
-	// Use password from input if available. This happens if we're restoring from
+	updateReq := v5.UpdateUserRequest{
+		Username: input.Role.StaticAccount.Username,
+	}
+	statements := v5.Statements{
+		Commands: input.Role.Statements.Rotation,
+	}
+
+	// Use credential from input if available. This happens if we're restoring from
 	// a WAL item or processing the rotation queue with an item that has a WAL
 	// associated with it
-	var newPassword string
 	if output.WALID != "" {
 		wal, err := b.findStaticWAL(ctx, s, output.WALID)
 		if err != nil {
-			return output, errwrap.Wrapf("error retrieving WAL entry: {{err}}", err)
+			return output, fmt.Errorf("error retrieving WAL entry: %w", err)
 		}
-
 		switch {
-		case wal != nil && wal.NewPassword != "":
-			newPassword = wal.NewPassword
+		case wal == nil:
+			b.Logger().Error("expected role to have WAL, but WAL not found in storage", "role", input.RoleName, "WAL ID", output.WALID)
+
+			// Generate a new WAL entry and credential
+			output.WALID = ""
+		case !wal.credentialIsSet():
+			b.Logger().Error("expected WAL to have a new credential set, but empty", "role", input.RoleName, "WAL ID", output.WALID)
+			if err := framework.DeleteWAL(ctx, s, output.WALID); err != nil {
+				b.Logger().Warn("failed to delete WAL with no new credential", "error", err, "WAL ID", output.WALID)
+			}
+
+			// Generate a new WAL entry and credential
+			output.WALID = ""
 		default:
```

Comment:
This is a bit of a refactor that improves readability IMO. The existing static role rotation code is pretty defensive (perhaps for a good reason, but it's not totally obvious). Happy to consider thoughts around how we can further improve readability here.

Reply:
This method is about 200 lines long and does many things. It might be more readable/maintainable to split it up into a few functions that do fewer things?

Reply:
I agree. I'd like to spend some time trying to reduce the complexity of this static rotation code in general. I'll take a look to see if there are any small, non-risky improvements that can be made as you've suggested. Otherwise, I'd prefer to leave any larger refactor out of this PR.

Reply:
Sounds good.